File: 2021/day9/day9.py
from collections import defaultdict


def part1(heightmap: list[list[int]]):
    low_points = []

    for i in range(1, len(heightmap) - 1):
        for j in range(1, len(heightmap[0]) - 1):
            if (heightmap[i][j - 1] > heightmap[i][j] and
                    heightmap[i][j + 1] > heightmap[i][j] and
                    heightmap[i - 1][j] > heightmap[i][j] and
                    heightmap[i + 1][j] > heightmap[i][j]):
                low_points.append(heightmap[i][j])

    print('Part 1:', sum(low_points) + len(low_points))


def part2(heightmap: list[list[int]]):
    visited = defaultdict(lambda: False)

    sizes = []

    for x in range(0, len(heightmap[0])):
        for y in range(0, len(heightmap)):
            size, visited = group_size(x, y, visited, heightmap)
            if size != 0:
                sizes.append(size)

    sizes = sorted(sizes)
    print('Part 2:', sizes[-1] * sizes[-2] * sizes[-3])


def group_size(x, y, visited, heightmap):
    size = 0
    if x in range(len(heightmap[0])) and y in range(len(heightmap)) and not visited[(x, y)] and heightmap[y][x] != 9:
        visited[(x, y)] = True
        size = 1
        right = group_size(x + 1, y, visited, heightmap)
        size += right[0]
        visited = right[1]
        left = group_size(x - 1, y, visited, heightmap)
        size += left[0]
        visited = left[1]
        down = group_size(x, y + 1, visited, heightmap)
        size += down[0]
        visited = down[1]
        up = group_size(x, y - 1, visited, heightmap)
        size += up[0]
        visited = up[1]
    return size, visited


if __name__ == '__main__':
    with open('in.txt') as file:
        lines = [[int(n) for n in list(line)] for line in file.read().splitlines()]

    width = len(lines[0])
    heightmap = []

    heightmap.append([9] * (width + 2))
    for line in lines:
        heightmap.append([9] + line + [9])
    heightmap.append([9] * (width + 2))

    part1(heightmap)
    part2(heightmap)

File: 2021/day9/day9.hs
module Main where

import Data.Foldable
import Data.List
import qualified Data.Map as Map
import System.IO

main :: IO ()
main = do
  contents <- readFile "in.txt"
  let lines' = map (map ((\digit -> read digit :: Int) . (: []))) $ lines contents
  let width = length (head lines')
  let heightmap = replicate (width + 2) 9 : map (\line -> 9 : line ++ [9]) lines' ++ [replicate (width + 2) 9]
  putStrLn $ "Part 1: " ++ show (part1 heightmap)
  putStrLn $ "Part 2: " ++ show (part2 heightmap)

part1 :: [[Int]] -> Int
part1 heightmap = sum lowPoints + length lowPoints
  where
    lowPoints =
      foldl
        ( \acc i ->
            acc
              ++ foldl
                ( \acc' j ->
                    if heightmap !! i !! (j - 1) > heightmap !! i !! j
                      && heightmap !! i !! (j + 1) > heightmap !! i !! j
                      && heightmap !! (i - 1) !! j > heightmap !! i !! j
                      && heightmap !! (i + 1) !! j > heightmap !! i !! j
                      then acc' ++ [heightmap !! i !! j]
                      else acc'
                )
                []
                [1 .. length (head heightmap) - 2]
        )
        []
        [1 .. length heightmap - 2]

part2 :: [[Int]] -> Int
part2 heightmap = product $ lastN' 3 sizes'
  where
    sizes' = nub $ sort $ fst sizes
    sizes =
      foldl
        ( \acc i ->
            case foldl
              ( \acc' j ->
                  case groupSize i j (snd acc') heightmap 0 "initial" of
                    (s, v) -> if s /= 0 then (s : fst acc', v `Map.union` snd acc') else (fst acc', v `Map.union` snd acc')
              )
              ([], Map.empty)
              [0 .. length heightmap -1] of
              (ac, v) -> (fst acc ++ ac, snd acc `Map.union` v)
        )
        ([], Map.empty)
        [0 .. length (head heightmap) -1]

groupSize :: Int -> Int -> Map.Map (Int, Int) Bool -> [[Int]] -> Int -> String -> (Int, Map.Map (Int, Int) Bool)
groupSize x y visited heightmap size direction
  | x `elem` [0 .. length (head heightmap) -1] && y `elem` [0 .. length heightmap -1] && Map.notMember (x, y) visited && heightmap !! y !! x /= 9 =
    case direction of
      "initial" ->
        case groupSize (x + 1) y (visited `Map.union` Map.fromList [((x, y), True)]) heightmap 1 "right" of
          (sr, vr) ->
            case groupSize (x - 1) y vr heightmap (1 + sr) "left" of
              (sl, vl) ->
                case groupSize x (y + 1) vl heightmap (1 + sr + sl) "down" of
                  (sd, vd) ->
                    case groupSize x (y - 1) vd heightmap (1 + sr + sl + sd) "up" of
                      (su, vu) -> (1 + sr + sl + sd + su, vu)
      _ -> groupSize x y visited heightmap size "initial"
  | otherwise = (0, visited)

lastN' :: Int -> [a] -> [a]
lastN' n xs = foldl' (const . drop 1) xs (drop n xs)